src/map/instance.cpp

```diff
@@ -257,3 +259,130 @@ bool CInstance::CheckFirstEntry(uint32 id)
     //insert returns a pair (iterator,inserted)
     return m_enteredChars.insert(id).second;
 }
+
+void CInstance::DisengageAll()
+{
+    for (EntityList_t::const_iterator it = m_charList.begin(); it != m_charList.end(); ++it)
+    {
+        CCharEntity* PCurrentChar = (CCharEntity*)it->second;
+
+        if(PCurrentChar->isAlive() && PCurrentChar->GetBattleTargetID() > 0)
+        {
+            PCurrentChar->PAI->Disengage();
+        }
+    }
+
+    for (EntityList_t::const_iterator it = m_allyList.begin(); it != m_allyList.end(); ++it)
+    {
+        CMobEntity* PCurrentAlly = (CMobEntity*)it->second;
+
+        if(PCurrentAlly->isAlive() && PCurrentAlly->GetBattleTargetID() > 0)
+        {
+            PCurrentAlly->PEnmityContainer->Clear();
+            PCurrentAlly->PAI->Disengage();
+            PCurrentAlly->PAI->Inactive(std::chrono::milliseconds(10000), false);
+        }
+    }
+
+    for (EntityList_t::const_iterator it = m_mobList.begin(); it != m_mobList.end(); ++it)
+    {
+        CMobEntity* PCurrentMob = (CMobEntity*)it->second;
+
+        if(PCurrentMob->isAlive() && PCurrentMob->GetBattleTargetID() > 0)
+        {
+            PCurrentMob->m_Aggro = false;
+            PCurrentMob->PEnmityContainer->Clear();
+            PCurrentMob->PAI->Disengage();
+            PCurrentMob->PAI->Inactive(std::chrono::milliseconds(10000), false);
+        }
+    }
+}
+
+void CInstance::StartAllyAssist(ALLY_ASSIST_MODE mode)
+{
+    // If there's no mobs, we can't assist. Don't waste time.
```

Comment:
why did you make this a core function?  instead of a mixin, or a generic lua function?